larq/optimizers_v2.py

```diff
@@ -1,10 +1,151 @@
+import warnings
+
 import tensorflow as tf
 import larq as lq
 
 from larq import utils
 from copy import deepcopy
 
 
+@utils.register_keras_custom_object
+class CaseOptimizer(tf.keras.optimizers.Optimizer):
+    """A case operation for optimizers that each train a subset of a model's variables.
+
+    An optimizer is used to train a variable iff its accompanying predicate evaluates to
+    `True`.
+
+    For each variable, at most one optimizer's predicate may evaluate to `True`. If no
+    optimizer's predicate evaluates to `True` for a variable, it is trained with the
+    `default` optimizer.
+
+    # Arguments
+    pred_opt_pairs: a list of `(tf.keras.optimizers.Optimzer, pred)` pairs, where `pred` 
+        takes one `tf.Variable` as argument and returns `True` if the optimizer should
+        train that variable, e.g. `pred(var) == True`.
+    default: a `tf.keras.optimizers.Optimizer` to be applied to any variable not claimed
+        by any other optimizer.
+    """
+
+    def __init__(self, pred_opt_pairs, default=None, name="optimizer_case"):
+        super().__init__(name=name)
+
+        # Type checks for (predicate, optimizer) pairs
+        for i, (predicate, optimizer) in enumerate(pred_opt_pairs):
+            if not callable(predicate):
+                raise TypeError(
+                    f"Expected callable predicate at `pred_opt_pairs[{i}][0]` but got `{type(predicate)}`."
+                )
+            if not isinstance(optimizer, tf.keras.optimizers.Optimizer):
+                raise TypeError(
+                    f"Expected `tf.keras.optimizers.Optimizer` at `pred_opt_pairs[{i}][1]` but got `{type(optimizer)}`."
+                )
+
+        # Type check for default optimizers
+        if default is not None and not isinstance(
+            default, tf.keras.optimizers.Optimizer
+        ):
+            raise TypeError(
+                f"Expected `tf.keras.optimizers.Optimizer` for `default` but got `{type(fallback_optimizer)}`."
+            )
+
+        self.pred_opt_pairs = pred_opt_pairs
+        self.default = default
+
+        self.var_opt_mapping = None
+        self.DEFAULT_OPT_INDEX = -1
+
+    def apply_gradients(self, grads_and_vars, name=None):
+        """Apply gradients to variables for each optimizer.
+        
+        On the first call to `apply_gradients()`, compute the mapping from variables to
+        optimizers and cache it in the `self.var_opt_mapping` dict for serialization and
+        faster access.
+        """
+
+        if self.var_opt_mapping is None:
+            grads_and_vars = list(grads_and_vars)
+            self._compute_var_opt_mapping(grads_and_vars)
+
+        train_ops = [
+            self._get_optimizer(var).apply_gradients([(grad, var)])
```

Comment:
I would prefer calling `apply_gradients` for each optimizer only once, due to performance reasons.

Reply:
So iterate over the `grads_and_vars` to sort them into separate lists for each optimizer first, and then passing these lists to each optimizer?

Reply:
Or we can switch the mapping, from `{var: optimizer}` to `{optmizer: var}`. Not sure what's better.

I'd prefer only call `apply_gradient` once per optimizer, since we would otherwise run [this code](https://github.com/tensorflow/tensorflow/blob/6977b8f7422b436c02aafde371716319149af542/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L407-L500) many times. Not sure if this is actually a performance problem, I'd need to look deeper into that.

Reply:
I tried `{optimizer: var}` but it ended up being a bit messy. Made some tweaks to now precompile lists of optimizers and `grads_and_vars` for each optimizer, so that `apply_gradients` only calls each optimizer once. See 9cd01e3.